### PR TITLE
Wait for pending ops to be saved before summarization

### PR DESCRIPTION
**Pull request by agarwal-navin**
## Description
There can be pending (unacked) ops in the summarizer client during summarization. This leads to eventually inconsistent and incorrect summaries. #16961 added failing summarization if there are pending ops **_after_** summary has been generated.
This change adds failing summarization if there are pending ops **_before_** generating summary. Additionally, it will fail for some time (default 1 second) for pending ops to be processed before starting to generate summary. The idea behind this change is that if there are pending ops before we even start to generate summary, it will fail anyway. So, might as well fail early and the wait is an optimization to give it a chance to start with a clean state.

## Key points
- The logic to wait for pending ops is behind a feature flag so that we can experiment with it and look at data before turning it on.
- Waiting for pending ops to be saved will happen with a timeout so that we don't end up waiting for a long time. The behavior when the timeout expires is to continue with summarization to keep the behavior same as today. Otherwise, it can lead to documents being stuck, i.e., not summarizing (maybe for a long time). We can tweak this behavior based on the data we see with the experiment.

[AB#4903](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/4903)

### CODE REVIEW

**Comment by agarwal-navin** [comment]
The second part of this solution AB#4904 is more important as per the data we see, i.e., if there are pending ops after summarization has completed, fail summarization and retry.
The changes in this PR are a subset of it and is an optimization - If we already know there are pending ops, summarization will fail later anyway, so might as well do something now. We can decide to just fail the summarization such that it is retried after a delay.

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 1]
```diff
@@ -3256,6 +3243,111 @@ export class ContainerRuntime
...
+		minimumSequenceNumber: number,
+		finalAttempt: boolean,
+		beforeSummaryGeneration: boolean,
+	): Promise<IBaseSummarizeResult | undefined> {
+		if (!this.hasPendingMessages()) {
+			return;
+		}
+
+		// If called before generating summary, wait for pending ops to be processed before proceeding.
+		if (beforeSummaryGeneration) {
```
Comment (tyler-cai-microsoft) [inline packages/runtime/container-runtime/src/containerRuntime.ts:3270]: The wait that occurs here isn't obvious from shouldFailOnPendingOps

Comment (tyler-cai-microsoft) [reply packages/runtime/container-runtime/src/containerRuntime.ts:3270]: It feels like this should be a separate function. I think duplicate this.hasPendingMessages() checks are better than this wait code being hidden, up to you. Feel free to resolve

**Review by tyler-cai-microsoft — APPROVED** [review]
LGTM - only minor comments

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 1 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/containerRuntime.ts:3270]: Yeah, good point. The waiting part is unexpected. I moved it to a separate function because `submitSummary` is already too big. Let me see what I can do.

Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/containerRuntime.ts]: Moved the waiting into the calling code.

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 2]
```diff
@@ -2965,6 +2971,51 @@ export class ContainerRuntime
...
+				const timeoutId = setTimeout(() => resolve(), pendingOpsTimeout);
+				this.once("saved", () => {
+					clearTimeout(timeoutId);
+					resolve();
+				});
+				this.once("dispose", () => {
+					clearTimeout(timeoutId);
+					reject(new Error("Runtime is disposed while summarizing"));
+				});
+			});
```
Comment (NicholasCouri) [inline packages/runtime/container-runtime/src/containerRuntime.ts:2994]: should we have a FG to bypass this code entirely ?

Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/containerRuntime.ts:2994]: The `this.validateSummaryBeforeUpload` is behind a feature flag. Do you mean just this await behind a FG?

Comment (NicholasCouri) [reply packages/runtime/container-runtime/src/containerRuntime.ts:2994]: I believe that would be enough :)

**Review by NicholasCouri — APPROVED** [review]

File: packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts [hunk 3]
```diff
@@ -228,15 +212,21 @@ async function waitForSummaryOp(container: IContainer): Promise<boolean> {
...
-describeNoCompat("Summarizer with local data stores", (getTestObjectProvider) => {
+describeNoCompat("Summarizer with local changes", (getTestObjectProvider) => {
 	let provider: ITestObjectProvider;
 
-	beforeEach(async () => {
+	beforeEach(async function () {
 		provider = getTestObjectProvider({ syncSummarizer: true });
+
+		// These tests validate client logic. Testing against multiple services won't make a difference.
+		if (provider.driver.type !== "local") {
```
Comment (markfields) [inline packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts:222]: I think there's a version of `describeNoCompat` that takes in a list of supported drivers.  Maybe it's only for `itExpects` though now that I think of it.  This is fine anyway, JFYI.

File: packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts [hunk 4]
```diff
@@ -172,7 +173,7 @@ const createContainer = async (
 	if (disableSummary) {
 		summaryConfigOverrides = { state: "disabled" };
 	} else {
-		const IdleDetectionTime = 100;
+		const IdleDetectionTime = 20;
```
Comment (markfields) [inline packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts:176]: What's the units here, ms?

File: packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts [hunk 5]
```diff
@@ -228,15 +212,21 @@ async function waitForSummaryOp(container: IContainer): Promise<boolean> {
...
 
-	beforeEach(async () => {
+	beforeEach(async function () {
 		provider = getTestObjectProvider({ syncSummarizer: true });
+
+		// These tests validate client logic. Testing against multiple services won't make a difference.
+		if (provider.driver.type !== "local") {
+			this.skip();
+		}
 		settings = [];
```
Comment (markfields) [inline packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts:225]: ?
```suggestion
		settings = {};
```

File: packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts [hunk 6]
```diff
@@ -256,7 +246,16 @@ describeNoCompat("Summarizer with local data stores", (getTestObjectProvider) =>
 				rootDataObject.containerRuntime,
 			);
 			rootDataObject._root.set("dataStore2", dataObject.handle);
-			const { summarizer } = await createSummarizer(provider, container);
+			const { summarizer } = await createSummarizerFromFactory(
```
Comment (markfields) [inline packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts:249]: Aside: Maybe `createSummarizerFromFactory` should use a params object to be self-documenting and more concise in the case of lots of `undefined`s

File: packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts [hunk 7]
```diff
@@ -313,6 +321,120 @@ describeNoCompat("Summarizer with local data stores", (getTestObjectProvider) =>
...
+			{
+				eventName: "fluid:telemetry:Summarizer:Running:Summarize_cancel",
+				clientType: "noninteractive/summarizer",
+				error: "PendingOpsWhileSummarizing",
+				beforeGenerate: true,
+			},
+		],
+		async () => {
+			// Wait for 100 ms for pending ops to be saved.
+			const pendingOpsTimeout = 100;
```
Comment (markfields) [inline packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts:336]: nit: `pendingOpsTimeoutMs` to include units in var name

File: packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts [hunk 8]
```diff
@@ -313,6 +321,120 @@ describeNoCompat("Summarizer with local data stores", (getTestObjectProvider) =>
...
+			const container1 = await provider.makeTestContainer();
+			const { summarizer, container: summarizerContainer } = await createSummarizer(
+				provider,
+				container1,
+				{ loaderProps: { configProvider: mockConfigProvider(settings) } },
+				undefined /* summaryVersion */,
+				mockLogger,
+			);
+
+			const defaultDataStore1 = await requestFluidObject<ITestDataObject>(
```
Comment (markfields) [inline packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts:349]: nit: entryPoint?

File: packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts [hunk 9]
```diff
@@ -313,6 +321,120 @@ describeNoCompat("Summarizer with local data stores", (getTestObjectProvider) =>
...
+				"expected PendingOpsWhileSummarizing",
+			);
+
+			// We should have received a PendingOpsWhileSummarizing event with all the pending ops not saved.
+			mockLogger.assertMatch([
+				{
+					eventName: "fluid:telemetry:Summarizer:Running:PendingOpsWhileSummarizing",
+					saved: false,
+					countBefore: pendingOpCount,
+					countAfter: pendingOpCount,
```
Comment (markfields) [inline packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts:381]: Put these in `details`?

File: packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts [hunk 10]
```diff
@@ -313,6 +321,120 @@ describeNoCompat("Summarizer with local data stores", (getTestObjectProvider) =>
...
+				undefined /* containerRuntimeFactoryType */,
+				registryStoreEntries,
+				mockLogger,
+				mockConfigProvider(settings),
+			);
+			await provider.ensureSynchronized();
+
+			await assert.rejects(
+				async () => {
+					await summarizeNow(summarizer);
```
Comment (markfields) [inline packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts:424]: How is this test exercising "ops are sent during summarize"?

File: packages/runtime/container-runtime/src/test/containerRuntime.spec.ts [hunk 11]
```diff
@@ -1465,5 +1513,172 @@ describe("Runtime", () => {
...
+				settings["Fluid.ContainerRuntime.Test.ValidateSummaryBeforeUpload"] = true;
+				containerRuntime = await ContainerRuntime.loadRuntime({
+					context: getMockContext(settings) as IContainerContext,
+					registryEntries: [],
+					existing: false,
+				});
+			});
+
+			it("summary is submitted successfully", async () => {
+				const summarizeResult = await containerRuntime.submitSummary({
```
Comment (markfields) [inline packages/runtime/container-runtime/src/test/containerRuntime.spec.ts:1539]: I'm amazed that this function is unit testable!

File: packages/runtime/container-runtime/src/test/containerRuntime.spec.ts [hunk 12]
```diff
@@ -1465,5 +1513,172 @@ describe("Runtime", () => {
...
+						localOpMetadata: unknown,
+						metadata: Record<string, unknown> | undefined,
+					): void;
+				};
+				const containerRuntimeWithSubmit =
+					containerRuntime as unknown as ContainerRuntimeWithSubmit;
+				// Submit a rejoin op and yield for it to be flushed from outbox to pending state manager.
+				containerRuntimeWithSubmit.submit(
+					{
+						type: ContainerMessageType.Rejoin,
```
Comment (markfields) [inline packages/runtime/container-runtime/src/test/containerRuntime.spec.ts:1636]: Idea - You could use compatDetails and a bogus type here too, to ensure no side effects.

Comment (markfields) [reply packages/runtime/container-runtime/src/test/containerRuntime.spec.ts:1636]: Actually don't do that, it wouldn't make sense for a local op.  Rejoin is probably tame enough anyway

File: packages/runtime/container-runtime/src/test/containerRuntime.spec.ts [hunk 13]
```diff
@@ -1465,5 +1513,172 @@ describe("Runtime", () => {
...
+						sequenceNumber: 0,
+						contents: {
+							type: ContainerMessageType.Rejoin,
+							contents: "something",
+						},
+					} as any as ISequencedDocumentMessage,
+					true /* local */,
+				);
+				// Advance the clock by the remaining time so that pending ops wait is completed.
+				clock.tick(1);
```
Comment (markfields) [inline packages/runtime/container-runtime/src/test/containerRuntime.spec.ts:1669]: Is this necessary?

File: packages/runtime/container-runtime/src/test/containerRuntime.spec.ts (hunk 12 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/test/containerRuntime.spec.ts:1636]: The only problem is that rejoin may be removed. But then compiler would complain and the test would be fixed.

File: packages/runtime/container-runtime/src/test/containerRuntime.spec.ts (hunk 13 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/test/containerRuntime.spec.ts:1669]: yeah, because the previoud tick in line 1654 did 1 ms less than the timeout.

File: packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts (hunk 4 above)
Comment (agarwal-navin) [reply packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts:176]: yes. updated the name to reflect it.

File: packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts (hunk 5 above)
Comment (agarwal-navin) [reply packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts:225]: haha yeah I noticed this too. It works apparently!

File: packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts (hunk 6 above)
Comment (agarwal-navin) [reply packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts:249]: Yeah that sounds good. Easier to change params too without changing a bunch of tests. Not doing in this PR though.

File: packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts (hunk 8 above)
Comment (agarwal-navin) [reply packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts:349]: Can't use it here because the test runtime factory doesn't pass initialize entry point yet.

File: packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts (hunk 10 above)
Comment (agarwal-navin) [reply packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts:424]: dataObject2 does that when summarizing by using `mixinSummaryHandler`. I will add a comment above to mention this.

File: packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts (hunk 9 above)
Comment (agarwal-navin) [reply packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts:381]: It's kinda hard to do this because not all sources of telemetry events support passing in details. For example, this event logs properties from [RetriableSummaryError](https://github.com/microsoft/FluidFramework/blob/74f413acc9a02e31c494a6cb3490d1684c9183a6/packages/runtime/container-runtime/src/summary/summaryGenerator.ts#L189) which does not accept it.
Also, I am not sure if mock logger supports comparing details type properties.

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 14]
```diff
@@ -3074,47 +3125,15 @@ export class ContainerRuntime
...
-						return {
-							stage: "base",
-							referenceSequenceNumber: summaryRefSeqNum,
-							minimumSequenceNumber,
-							error,
-						};
-					}
+				const pendingMessagesFailResult = await this.shouldFailSummaryOnPendingOps(
+					summaryNumberLogger,
+					this.deltaManager.lastSequenceNumber,
```
Comment (markfields) [inline packages/runtime/container-runtime/src/containerRuntime.ts:3130]: referenceSequenceNumber in the result used to be set to `summaryRefSeqNum`, why the change?

Looking above, I guess they'll be the same anyway, as of whenever `checkContinue` was called, and DM is paused so should be the same.

Right?

File: packages/runtime/container-runtime/src/test/containerRuntime.spec.ts (hunk 13 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/test/containerRuntime.spec.ts:1669]: But why do we need to hit the timeout?  Just to prove that it won't actually fire, since the last (only) pending op was processed?

File: packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts (hunk 9 above)
Comment (markfields) [reply packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts:381]: k

File: packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts (hunk 8 above)
Comment (markfields) [reply packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts:349]: k. I can't keep track of where it is and isn't present yet in the test infra so sorry if I keep asking :P

File: packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts (hunk 5 above)
Comment (markfields) [reply packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts]: haha yeah JS for the win

**Review by markfields — APPROVED** [review]
LGTM. Let's get the plan for how we want to stage all these feature gates nailed down - I'll go review what you shared the other day now

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 14 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/containerRuntime.ts:3130]: Good catch. I copied the function call from where its called above. There `summaryRefSeqNum` is not set so I passed in lastSequenceNumber directly. Here, its better to use `summaryRefSeqNum` to be consistent.

File: packages/runtime/container-runtime/src/test/containerRuntime.spec.ts (hunk 11 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/test/containerRuntime.spec.ts:1539]: I wish that function was simpler though. Hopefully, it will be once all our planned work is done.

File: packages/runtime/container-runtime/src/test/containerRuntime.spec.ts (hunk 13 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/test/containerRuntime.spec.ts:1669]: Not sure if I missing soemthing but if the timeout doesn't fire, the submitSummary call won't complete right? These tests use fake timers so we have to manually advance them otherwise submitSummary will be stuck waiting for pending ops.

File: packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts (hunk 8 above)
Comment (agarwal-navin) [reply packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts:349]: Haha yeah its confusing. I would honestly not worry about it here and do all of them when request pattern is completely removed. The tests will have to be updated then.

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 14 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/containerRuntime.ts]: Fixed it.

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 15]
```diff
@@ -3252,6 +3271,72 @@ export class ContainerRuntime
...
+			const error = new RetriableSummaryError(
+				"PendingOpsWhileSummarizing",
+				retryDelayMs / 1000,
+				{
+					count: this.pendingMessagesCount,
+					beforeGenerate: beforeSummaryGeneration,
+				},
+			);
+			return {
+				stage: "base",
```
Comment (markfields) [inline packages/runtime/container-runtime/src/containerRuntime.ts:3332]: Do you have test coverage of this when the validateBeforeUpload setting is false?

Comment (markfields) [reply packages/runtime/container-runtime/src/containerRuntime.ts:3332]: Doesn't look like it - let's add that test, or hardcode a dependency between those settings (so pendingOps enforcement can only be enabled if validateBeforeUpload is enabled)

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 16]
```diff
@@ -3252,6 +3271,72 @@ export class ContainerRuntime
...
+		if (!this.hasPendingMessages()) {
+			return;
+		}
+
+		// If "SkipFailingIncorrectSummary" option is true, don't fail the summary in the last attempt.
+		// This is a fallback to make progress in documents where there are consistently pending ops in
+		// the summarizer.
+		if (
+			finalAttempt &&
+			this.mc.config.getBoolean("Fluid.Summarizer.SkipFailingIncorrectSummary")
```
Comment (markfields) [inline packages/runtime/container-runtime/src/containerRuntime.ts:3300]: You're going to switch this to default to "skip" right?

File: packages/runtime/container-runtime/src/test/containerRuntime.spec.ts (hunk 13 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/test/containerRuntime.spec.ts:1669]: The Promise where the Timeout is also has this:

```ts
this.once("saved", () => {
					clearTimeout(timeoutId);
					resolve();
				});
```

Shouldn't processing that pending op trigger "saved" event?  I didn't track through the code to verify that, but that's what I thought that event was about.  In that case the promise will resolve even if the timeout didn't yet fire.
